### ISSUE TITLE
CNI repair broken for non-restartable pods

### ISSUE THREAD

**Issue by howardjohn** [istio/istio#49210]
When a job has restartPolicy=Never or low backoffLimit, CNI race repair may not work. This is because the pod will not just crashloop until its fixed, it will stop immediately.

This applies even to the new `repairPods` mode, not just label/delete pods mode, because even `repairPods` is triggered by the pod crashlooping.

**Comment by bleggett** [comment]
Thinking about it, I guess this is not practically fixable except by adopting 

https://github.com/istio/istio/pull/48818 / https://github.com/istio/istio/pull/49092

to entirely obviate the need for crash-based repair.

**Comment by bleggett** [comment]
Now that the untaint controller is in I would actually advocate we drop this form of race repair entirely, and point people to the untaint controller instead.

Need to do a bit of work there still tho.

**Comment by howardjohn** [comment]
> Encountered the same problem. The most confusing part was that repair with `deletePods: true` also didn't work. I understand why network repairing won't be effective without restarts, but deleting the pod should work perfectly fine.
> 
> Turns out, the code always checks `LastTerminationState` struct to confirm the pod is "broken", which is only present if container has restarted, even though there's absolutely no logical reason for that requirement when the repair mode is configured to delete pods. https://github.com/istio/istio/blob/1.24.3/cni/pkg/repair/repaircontroller.go#L263

The reason is it is how we identify which pods to repair

**Comment by bleggett** [comment]
> > Encountered the same problem. The most confusing part was that repair with `deletePods: true` also didn't work. I understand why network repairing won't be effective without restarts, but deleting the pod should work perfectly fine.
> > Turns out, the code always checks `LastTerminationState` struct to confirm the pod is "broken", which is only present if container has restarted, even though there's absolutely no logical reason for that requirement when the repair mode is configured to delete pods. https://github.com/istio/istio/blob/1.24.3/cni/pkg/repair/repaircontroller.go#L263
> 
> The reason is it is how we identify which pods to repair

Yeah but they're right tho - the identification will always fail in some cases, so either we need to drop the clause or make it conditional in some way.

**Comment by artem-nefedov** [comment]
> > Encountered the same problem. The most confusing part was that repair with `deletePods: true` also didn't work. I understand why network repairing won't be effective without restarts, but deleting the pod should work perfectly fine.
> > Turns out, the code always checks `LastTerminationState` struct to confirm the pod is "broken", which is only present if container has restarted, even though there's absolutely no logical reason for that requirement when the repair mode is configured to delete pods. https://github.com/istio/istio/blob/1.24.3/cni/pkg/repair/repaircontroller.go#L263
> 
> The reason is it is how we identify which pods to repair

When the pod has `restartPolicy: Never`, it would've been enough to run the same checks (exit code/message) on the state from `container.State.Terminated` to confirm that the pod is broken.

**Comment by howardjohn** [comment]
> > > Encountered the same problem. The most confusing part was that repair with `deletePods: true` also didn't work. I understand why network repairing won't be effective without restarts, but deleting the pod should work perfectly fine.
> > > Turns out, the code always checks `LastTerminationState` struct to confirm the pod is "broken", which is only present if container has restarted, even though there's absolutely no logical reason for that requirement when the repair mode is configured to delete pods. https://github.com/istio/istio/blob/1.24.3/cni/pkg/repair/repaircontroller.go#L263
> > 
> > 
> > The reason is it is how we identify which pods to repair
> 
> When the pod has `restartPolicy: Never`, it would've been enough to run the same checks (exit code/message) on the state from `container.State.Terminated` to confirm that the pod is broken.

Ah gotcha, didn't realize we were talking about state.terminated vs LastTerminationState, I thought you meant the general idea of looking at an exit code. 

Seems like that may work for `delete` mode (though not repair)

**Comment by jbilliau-rcd** [comment]
Hey @howardjohn , glad I found this; we are experiencing this very issue with using `k6` to run load tests on in-cluster applications. If the amount of k6 pods required nodes to spin up(99% of the time, they do), the pods try to come up before the istio-cni, can't, and then just fail permanently _(this does not happen with `deployment` or `rollout` created pods, since as you say in the original post, they are restartable)_

Is the suggested solution this other thread I found on using a new "untaint" controller? If we can ensure all of our nodes are tainted at creation, are there any cons to switching to this from the existing default of `repairPods`? _(which has been working for us just fine until this k6 situation came along)_

https://github.com/istio/istio/pull/48818
https://github.com/istio/istio/issues/51176
